Modules/Module_Import/Import/xASL_imp_DetermineStructureFromSourcedata.m

```diff
@@ -240,42 +258,38 @@ function xASL_imp_PrintRun(x,thisSubject,thisVisit,thisRun)
     %% Preallocate space for (global) counts
     [x.importOverview.(sFieldName),x.importOverview.(sFieldName).(vFieldName)] = ...
         xASL_imp_PreallocateGlobalCounts(x.modules.import.nSubjects,x.importOverview.(sFieldName),x.importOverview.(sFieldName).(vFieldName));
-    
 
 end
 
-
+% -----------------------------------------------------------------
 %% Add fields of this subject/visit
+% -----------------------------------------------------------------
 function x = xASL_imp_thisSubjectVisit(x,sFieldName,vVisitIDs,vFieldName)
 
-
     % Visits (= sessions in BIDS)
     x = xASL_imp_AddVisitNames(x,sFieldName);
-    
-    
+   
     % Sessions (= run in BIDS)
     x = xASL_imp_AddSessionNames(x,sFieldName,vFieldName,vVisitIDs);
     
-    
     % Scans (= actual DICOM data)
     x = xASL_imp_AddScanNames(x,sFieldName,vFieldName,vVisitIDs);
-    
 
     % Sessions (= runs in BIDS)
     x = xASL_imp_AddSessions(x,sFieldName,vFieldName);
-
-
 end
 
-
+% -----------------------------------------------------------------
 %% Add visit names
+% -----------------------------------------------------------------
 function x = xASL_imp_AddVisitNames(x, sFieldName)
 
     if isempty(x.modules.import.imPar.visitNames)
         if isempty(x.importOverview.(sFieldName).visitIDs)
+			% In case that visit names are not filled, we name the visit by their number
```

Comment:
What happens here if some fields are empty and others are not? Then you get some visits having an index number and other visits having the original alphanumeric string. I know that this should not happen because of the unique function above but always better to have a check (I don't mind if the check is here or above).

Reply:
I agree. And I have thought of this option. That's why we copy the original alphanumeric string, and in case it's empty we provide a number, but as a string. So in the end, all values are strings and thus have a consistent type.

So a check shouldn't be necessary.

Reply:
OK so consider the following (with empty visittoken/no visit token specified):

/sourcedata/Subject001_20230101/
/sourcedata/Subject001_20240201/
/sourcedata/Subject001_/

What will happen? Do we give an error above once one of the unique visits is empty? Or a warning and we assign 'missingSessionValue'?

Probably the second is nicer for the user, because rawdata is more user-readable than our huge log files, and the huge sourcedata folder. In this case, it would become:

/rawdata/sub-001/20230101/
/rawdata/sub-001/20240101/
/rawdata/sub-001/missingSessionValue/

So I would go for the warning and replacement of empty alphanumerics by "missingSessionValue". In the case that no visit token is specified. In fact, we could do the same when visit tokens are specified, e.g. if we have specified this as visit tokens it would have been something like:

visitTokens = {'^20230101$', '1', '^20240201$', '2', '^20240401$', '3'}

and one would still get the above situation. So the replacement of empty visit values could go here or quite down in the code.

But now I actually think that Paul's findRegExp function would not detect visits with empty values, unless we change the regular expression to also except empty (which is easy, it would just be (| or |). ).
The same goes for runs. It would be good to add such functionality if it doesn't take too much to program (it remains of course the user's responsibility).

Reply:
1. If folderhierarchy token regexp is (Subject\d(3))_(\d(8)) then it will simply not even find it. 

2. So I guess you are refering to the case of having (Subject\d(3))_(.*) which will find it and return an empty token. In case we have defined visitTokenAliases in sourcestructure.json, we will have to explicitly define an alias for this empty token - this already works. I would ignore this in case this alias doesn't fall within known aliases because what we will do in case we have several unknown token - I would skip all unknown tokens.

3. So the only relevant option that we should address is token renaming and an empty token, for which we can indeed create a "missingSessionValue". Done in the last commit.

Reply:
Let me try to explain again.

By "no visit token specified" I indeed meant something like 
folderHierarchy : (Subject\d(3))_(.*)
and no row that says visitTokenAliases

the sourcedata is:

/sourcedata/Subject001_20230101/
/sourcedata/Subject001_20240201/
/sourcedata/Subject001_/

So we need to fix two things:

1.  If you have no visitTokenAliases, you create a dummy visitTokenAlias for each unique value right?
So you would create this for '20230101' and '20240201' but not for ''. So you need to add '' if this is not too much to reprogram (like I describe above but some reason you didn't read? 

example dummy visitTokenAliases created internally if visitTokenAliases was not specified in sourceStructure.json in this example you probably create:

visitTokenAliases = {'^20230101$', '1', '^20240201$', '2'}

but it should create:

visitTokenAliases = {'^20230101$', '1', '^20240201$', '2','^$','missingSessionValue}

So it should always add this.






3. No, I'm first referring to the case where you have NO visit token provided, and

Reply:
You have to explain it 3 times because you don't read things properly :)

I understood it already the first time and did exactly what you asked for and wrote that I did that - unspecified/missing visit tokens were renamed to "missingSessionToken" sessions.

You were right that some extra edits were needed, but that was just because I didn't have time to test things, not because I wouldn't answer everything.

Reply:
I'm not saying you understood things, I'm saying that if you don't communicate clearly to me what was done why, how am I supposed to check if I agree within reasonable time investment (i.e. without reading all the code? ;)

So: we have 2 situations, A) one with visitTokenAliases defined, B) one without being defined. We want to support not defining visitTokenAliases.

A) So you argue that if visitTokenAliases are explicitly defined, it would skip those that slightly differ or are empty. That's fine perhaps, because the user defining these tokens will probably have more control and check better than the user not defining visitTokenAliases, so we don't need to "help" the user by also converting the odd ones.
B) If you created a dummy visitTokenAliases from all visits, the empty ones need replacing by "missingSessionValue"


We might want to explain clearly that either A) all visit visitTokenAliases are explicitly defined, or B) visitTokenAliases are not defined at all. But that there is no in between.


Note also that the comment above is confusing, it says that for missing visit names, we will replace them by their number. Is this comment now incorrect, or is this part still replacing missing visit names?

i.e., does this code does:

A) convert
visitTokenAliases = {'^visit_1', 1, '^visit_2'}
to
visitTokenAliases = {'^visit_1', 1, '^visit_2', 2}

OR does this code does
B) create a dummy list of visitTokenAliases when visitTokenAliases was not provided?

Reply:
> So: we have 2 situations, A) one with visitTokenAliases defined, B) one without being defined. We want to support not defining visitTokenAliases.
Yes. This was the original goal defined in the issue.

> A) So you argue that if visitTokenAliases are explicitly defined, it would skip those that slightly differ or are empty. That's fine perhaps, because the user defining these tokens will probably have more control and check better than the user not defining visitTokenAliases, so we don't need to "help" the user by also converting the odd ones. B) If you created a dummy visitTokenAliases from all visits, the empty ones need replacing by "missingSessionValue"

Following your previous comments, I have already implemented the following:
1) If tokens are defined, it only takes those that are defined and ignore those that are undefined - and report a warning.
2) If tokens are not defined, it copies them exactly, but removes non-alphanumerical characters.
3) Following your first comment, I have implemented that for non-defined tokens, it additionally converts a session with an empty token to "missingSessionValue" - e.g. case 2 above. Following your second comment, I have broaden the replacing the empty session by "missingSessionValue" for those cases where tokenVisitAliases are defined. If you want to remove this replacement for provided tokenVisitAliases, I can easily do that.

> We might want to explain clearly that either A) all visit visitTokenAliases are explicitly defined, or B) visitTokenAliases are not defined at all. But that there is no in between.

This will be added to the documentation before merging the issue. We have a task for this in the issue.

> Note also that the comment above is confusing, it says that for missing visit names, we will replace them by their number. Is this comment now incorrect, or is this part still replacing missing visit names?

This is not really confusing. We have three cases.
1) TokenvisitAliases defined
2) TokenVisitAlises undefined, but tokenVisit defined
3) TokenVisit undefined.
For case 3, where we don't really even attempty to extract tokens for visit (token order = [1 0 2 3]), we always add visit _1 to each subject. That's something we added in previous issues on your request.

> i.e., does this code does:
> 
> A) convert visitTokenAliases = {'^visit_1', 1, '^visit_2'} to visitTokenAliases = {'^visit_1', 1, '^visit_2', 2}
> OR does this code does B) create a dummy list of visitTokenAliases when visitTokenAliases was not provided?

The code does neither - it doesn't work with tokenAliases, because no token was extracted.

Reply:
This whole section was removed as completely useless. You can resolve.